Support scene activation with multiple configured hue bridges

alistairg opened the pull request:
## Description:

When using multiple hue bridges, the 'hue_activate_scene' service would really only activate a scene on the last-most loaded bridge.

This change:

- Moves the service registration to the overall component
- When the hue_activate_scene service is called, all bridges are asked to activate the scene - but not refresh their scene and group cache if the request fails (on that bridge)
- If any one bridge was able to activate the scene, the service call completes successfully
- If no bridges activated the scene a second, more expensive, pass is made asking each bridge to refresh their scene and group cache and retry
- If still no bridge activated a scene, an error is output to the log

Emergent functionality:

- If multiple bridges have the same group name and scene name, the scene will be recalled on each matching bridge. This is reasonably expected behaviour.

**Related issue (if applicable):** fixes #16452

**Pull request in [home-assistant.io](https://github.com/home-assistant/home-assistant.io) with documentation (if applicable):** home-assistant/home-assistant.io#<home-assistant.io PR number goes here>

## Example entry for `configuration.yaml` (if applicable):
```yaml

```

## Checklist:
  - [X ] The code change is tested and works locally.
  - [ X] Local tests pass with `tox`. **Your PR cannot be merged unless tests pass**
  - [ X] There is no commented out code in this PR.

If user exposed functionality or configuration variables are added/changed:
  - [ ] Documentation added/updated in [home-assistant.io](https://github.com/home-assistant/home-assistant.io)

If the code communicates with devices, web services, or third-party tools:
  - [ ] New dependencies have been added to the `REQUIREMENTS` variable ([example][ex-requir]).
  - [ ] New dependencies are only imported inside functions that use them ([example][ex-import]).
  - [ ] New or updated dependencies have been added to `requirements_all.txt` by running `script/gen_requirements_all.py`.
  - [ ] New files were added to `.coveragerc`.

If the code does not interact with devices:
  - [ ] Tests have been added to verify that the new code works.

[ex-requir]: https://github.com/home-assistant/home-assistant/blob/dev/homeassistant/components/keyboard.py#L14
[ex-import]: https://github.com/home-assistant/home-assistant/blob/dev/homeassistant/components/keyboard.py#L54

syssi committed:
Refactor motion sensor

syssi committed:
Improve debug output

larsvinc committed:
Added support for spot cleaning that was introduced in pybotvac 0.12.

larsvinc committed:
Corrected formating.

larsvinc committed:
Added missing operator.

alistairg committed:
Support scene activation with multiple configured hue bridges

alistairg committed:
Shortened line lengths per code review

alistairg committed:
Shortened variable names

thomasloven committed:
Allow other icon prefixes than mdi: (#19872)

exxamalte committed:
Geo Location -> Geolocation (class names and unit test comments) (#19877)

* fixed geolocation naming in class names

* fixed geolocation naming in comments in unit test

alistairg committed:
Resolves #17196, Resolves #18739 - Hue Beyond light fixture errors (#19874)

* Resolves #17196, Resolves #18739 - Hue Beyond light fixtures being incorrectly recognized

* Removed long code lines that were failing code review

* Removed trailing whitespace

pvizeli committed:
Update OZW to 0.1.2 (#19878)

* Update ozw 0.1.2

* Update requirements_all.txt

Rendili committed:
add entity support to hive (#19879)

balloob commented on homeassistant/components/hue/__init__.py:
```diff
@@ -53,14 +54,47 @@
+
+        # Call the set scene function on each bridge
+        results = []
+        for dummy_host, bridge in hass.data[DOMAIN].items():
```
```suggestion
        for bridge in hass.data[DOMAIN].values():
```

balloob commented on homeassistant/components/hue/__init__.py:
```diff
@@ -53,14 +54,47 @@
+        # Call the set scene function on each bridge
+        results = []
+        for dummy_host, bridge in hass.data[DOMAIN].items():
+            result = await bridge.hue_activate_scene(call,
```
Collect all coroutines in a list and then await that list of coroutines, so it happens in parallel.

```python
tasks = [bridge.hue_activate_scene(…) for bridge in hass.data[DOMAIN].values()]

results = await asyncio.gather(tasks)

if not any(result is False for result in results):
    return

if skip_reload:
    …
```

alistairg replied:
Great suggestion - thanks. Still getting used to asyncio ;p

FlorianLudwig committed:
assign user to websocket connection when using legacy_api_password (#19797)

rohankapoorcom committed:
Migrate geofency over to the Webhook component (#18951)

* Migrate geofency over to the Webhook component

* Return web.Response correctly

* Fix test

* Lint

* Fix error that tests caught

balloob committed:
Refactor motion sensor of the xiaomi_aqara platform (#19805)

* Refactor motion sensor

* Improve debug output

balloob committed:
Fix deprecation warning (#19882)

Rendili committed:
Bug fix with getting a device battery level when API unavailable for Hive (#19841)

* hive updates - bug fix and add entity registration

* remove hive entity registration code

balloob committed:
Updated frontend to 20190109.0

balloob committed:
Update translations

alistairg committed:
Improved call to set_scene with parallel execution on multiple hubs

alistairg committed:
Added additional sanity checking

Installations with legacy configuration for Hue may run into issues as they track objects other than just bridges in config.

This change adds further sanity checking to ensure that any non-bridge objects are ignored.

bachya committed:
Bump pyflunearyou to 1.0.1 (#19899)

tokarev-artem committed:
Removed mkdir, If the WORKDIR doesn’t exist, it will be created. (#19892)

RomRider committed:
min_max sensor support for STATE_UNAVAILABLE

royduin committed:
Plant monitor defaults (#19891)

* Plant monitor defaults

* houndci-bot fixes

mitchellrj committed:
Correctly map Nest hvac_state to Home Assistant states. (#19895)

Adminiuga committed:
Don't map LevelControl to light for single cluster devices. (#19929)

cliffordwhansen committed:
Add btle_name attribute to devices (#19915)

* Update googlehome.py

Added name from bluetooth device to attributes

* Update homeassistant/components/device_tracker/googlehome.py

Check if key exists before assigning name

Co-Authored-By: cliffordwhansen <clifford@nighthawk.co.za>

fabaff committed:
Upgrade youtube_dl to 2019.01.10

dshokouhi committed:
Fix botvac connected alert retrieval

abmantis committed:
Add Alexa's EndpointHealth reporting (#19784)

* add Health reports

* add health report for all devices

* update tests

* Update homeassistant/components/alexa/smart_home.py

Co-Authored-By: abmantis <abmantis@users.noreply.github.com>

* lint

* add tests

nickneos committed:
Remove Discovery dependency from konnected.py (#19910)

alistairg committed:
Added additional comments and shortened lines per code review

alistairg committed:
Removed trailing whitespace

balloob reviewed (approved):
Ok to merge when build passes 🎉

arigilder committed:
Add sensors to jewish_calendar for upcoming Shabbat times (#19278)

* Initial pass of cleanup for shabbat_times

* Switch to async defs

* First pass of unit tests + fixture data

* Completion of first round of unit tests, 100% passing

* Unit tests for state restoring

* Style fixes

* More style fixes

* Lint fix

* Add upcoming candelighting and havdalah sensors

* Add unit tests, remove havdalah offset

* More unit tests + small bugfix for weekly_portion

* Add issur melacha sensor

* Remove old shabbat_times work-in-progress files

* Bump required version of hdate

* Add havdalah offset config parameter

* Bump hdate version required

* Pin hdate requirement

* Lint fixes

* Changes based on review + API changes for hdate 0.8.7

* Add three-day holiday unit tests

* Remove debugging line

* Add missing docstring

* Fix doc lint comment

SNoof85 committed:
add_entities -> async_add_entities (#19943)

Better state handling at HA startup

rymsha committed:
Add nad telnet (#19704)

* fix cla-bot

* fix bug introduced after linter complaint

* merge two components into one
support telnet port configuration

* remove obsolete nadtcp component. nad component must be used instead.

* back to correct nad_receiver version

fronzbot committed:
Upgrade blinkpy and use calibrated temperature for sensor (#19723)

abmantis committed:
Small refactoring for the alexa component (#19782)

* small refactoring

* fix tests

fabaff committed:
Upgrade pylast to 3.0.0 (#19938)

balloob committed:
Fix botvac connected alert retrieval (#19937)

therve committed:
Remove spurious libzwave error (#19928)

If a network_key is not configuired, the following error is logged:
TypeError: expected bytes, NoneType found
Exception ignored in: 'libopenzwave.str_to_cppstr'
TypeError: expected bytes, NoneType found

We don't need to set the key if it's None, let's skip in that case.

Commit:
Add emulated_roku component (#17596)

* Add emulated_roku component

* Add emulated_roku config tests

* Fix emulated_roku test dependencies

* Remove emulated_roku yaml support, add tests

* Add yaml support, simplify config flow

* Improve emulated_roku code quality

* Fix emulated_roku translation, improve code quality

* Fix emulated_roku translation

* Bump emulated_roku to 0.1.6 to fix SSDP discovery

* Bump emulated roku to 0.1.7, refactor component start/stop methods

Commit:
Change state() to try/except to catch KeyError (#19935)

* Change state() to try/except to catch KeyError

When Tautulli is up but Plex is down, the API doesn't return a 'stream_count' key. This causes calls to state() to raise KeyError exceptions. The new code includes a try/except to catch the KeyError and return -1 signifying that the Tautulli API cannot talk to Plex

* Update tautulli.py

awarecan committed:
Only authenticate request when owner can be found

fabaff committed:
Upgrade mutagen to 1.42.0 (#19956)

fabaff committed:
Upgrade pillow to 5.4.1 (#19958)

fabaff committed:
Upgrade ruamel.yaml to 0.15.85 (#19960)

fabaff committed:
Upgrade pysnmp to 4.4.8 (#19961)

fabaff committed:
Upgrade keyring to 17.1.1 (#19962)

soldag committed:
Add support for restoring state to rpi_gpio_pwm (#19944)

balloob committed:
min_max sensor support for STATE_UNAVAILABLE (#19914)

eliseomartelli committed:
add service type in name (#19980)

balloob committed:
Add support for spot cleaning that was introduced in pybotvac 0.12 (#19857)

* Added support for spot cleaning that was introduced in pybotvac 0.12.

* Corrected formating.

* Added missing operator.

balloob committed:
Merge pull request #19959 from home-assistant/awarecan-trusted-network-auth-must-have-user

Trusted Network Auth: only authenticate request when owner can be found

filcole commented:
I think this is going to cause problems if trying to activate the special all lights "hidden" scene.  Should that scene be triggered on both hubs?  See https://github.com/home-assistant/home-assistant/pull/8663 and https://github.com/home-assistant/home-assistant/issues/8652 for details of the special group.

syssi committed:
Support next generation of the Xiaomi Mi Smart Plug (chuangmi.plug.m3) (#19972)

* Support next generation of the Xiaomi Mi Smart Plug (chuangmi.plug.m3)

* Fix indent

jarlehjortland committed:
When tradfri experience communication errors make the lights/devices unavailable. (#19288)

balloob committed:
Allow embedded platforms (#19948)

* Allow embedded platforms

* Fix test

dmulcahey committed:
Add services and helper functions to support a config panel for ZHA (#19664)

* reconfigure zha device service

add log line to reconfigure service for consistency

* add entity functions to support new services

* added new services and web socket api and split them into their own module

* support manufacturer code

logging to debug

get safe value for manufacturer

* update services.yaml

* add comma back

* update coveragerc

* remove blank line

* fix type

* api cleanup - review comments

* move static method to helpers - review comment

* convert reconfigure service to websocket command - review comment

* change path

* fix attribute

balloob committed:
Updated frontend to 20190109.1

balloob committed:
Fix warning (#19946)

* Fix warning

* Update service.py

balloob committed:
Fix warning (#19946)

* Fix warning

* Update service.py

grea09 committed:
Adding IPv6 to fail2ban sensor (#19457)

* Fixing fail2ban regex for ipv6

* Adding IPv6 tests for fail2ban

* Formating code for hound

* Formating again

* Formating again 2

mikeage committed:
Always use datetime and timedelta in camera.proxy instead of int/float (#19571)

Commit:
Include Scripts/ directory to .gitignore - this is created by virtualenv on Windows (#18918)

MartinHjelmare committed:
Add mysensors state update delay (#18891)

* Add mysensors state update delay

* Schedule state update after delay to avoid updating state multiple
  times during the same short timespan.

* Code review

oblogic7 committed:
Add ability to monitor relay events (#18730)

* Add ability to monitor relay events

* Account for empty events array instead of none.

fredrike committed:
Add support for 'via_hub' for device_info (#19454)

* Add support for 'via_hub'

* Update config schema

* add domain to via_hub

* add tests for via_hub

achatham committed:
Fix the anthemav component by removing a debugging line. (#19979)

This lane ended up calling vars(transport) on an asyncio Transport
object. In a standard setup, that's a python object provided by syncio,
and it works. Home Assistant injects uvloop into asyncio, which makes this
a Python C object, and those don't support vars().

alistairg committed:
Merge branch 'hue-scene' of https://github.com/alistairg/home-assistant into hue-scene

alistairg commented:
I managed to hork this codeline. Sorry. Closing this PR - will fix and resubmit.

balloob commented:
No problem. Please tag me so I am sub'd to the new PR.
